src/utils/hooks/useListTemplates.tsx

```diff
@@ -17,30 +37,96 @@ const useListTemplates = (templatePermissions: TemplatePermissions, isLoading: b
       setLoading(false)
     }
 
-    if (data && data?.templates?.nodes) {
-      const allTemplates = data?.templates?.nodes as Template[]
-      const filteredTemplates: TemplatesDetails = []
-      allTemplates.forEach(({ code, name }) => {
-        const permissionFound = Object.entries(templatePermissions).find(([key]) => key === code)
-        if (permissionFound) {
-          const [key, permissions] = permissionFound
-          filteredTemplates.push({
-            name: name as string,
-            code,
-            permissions: permissions,
-          })
-        }
+    if (data?.templates?.nodes) {
+      const filteredTemplates = (data?.templates?.nodes || []).filter(
+        (template) => templatePermissions[String(template?.code)]
+      ) as Template[]
+      if (filteredTemplates.length === 0) return setLoading(false)
+
+      const templates = filteredTemplates.map((template) =>
+        convertFromTemplateToTemplateDetails(template, templatePermissions)
+      )
+
+      setTemplatesData({
+        templates,
+        templatesByCategory: getTemplatesByCategory(templates),
       })
```

Comment:
Kind of a big change - but it seems a little more straight-forward to me. You could do in the same loop a separation of templates per category in a object like this:
```
[
   {
      category: TemplateCategoryDetails
      templateList: TemplateInList[]
   },
   ...
]
```

And perhaps the category.title and category.icon doesn't need to go inside the `TemplateInList` also, just removing it from the type. I'll continue looking with that in mind see if there is any flaws.

Reply:
Hmm, prefer separating calculations based on their result, similar to how the progress calculations are done for review vs application. Similar in the review structure calculation, could potentially have a big loops calculating isPendingReview, isNewApplicationResponse, isActiveReviewResponse, thisReviewLatest response.. etc..)  -> but I think it's easier to digest logic when it's separated this way (and easier to find and change pieces of logic). It does create redundancy with space (having templateCategory in the TemplateInList) and it's not optimised (having to loop over the same element multiple times). I prefer readability over premature optimisation though. But maybe it's not as readable as i think ?

Reply:
Yes, to me the double loop is compromising to quickly read. It would be nice to have all in a big loop, but I don't quite see the connection between this one and the deduction of fields in the progress... So I didn't get the correlation, if you don't want to change, it's fine. I was just pointing out something that could be improved from my perspective.